File: benchmarks/regex-redux/regex_redux.js
/**
 * The Computer Language Benchmarks Game
 * https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 *
 * regex-redux benchmark in JavaScript
 */

'use strict';

const fs = require('fs');

function main() {
  // Read input
  let input = '';
  const buffer = Buffer.alloc(65536);
  let bytesRead;
  
  while ((bytesRead = fs.readSync(0, buffer, 0, buffer.length)) > 0) {
    input += buffer.slice(0, bytesRead).toString('utf8');
  }
  
  // Remove header and newlines
  const sequence = input.replace(/>.*\n|\n/g, '');
  
  // Count patterns
  const initialLength = input.length;
  const codeLength = sequence.length;
  
  const patterns = [
    { regex: /agggtaaa|tttaccct/ig, name: 'agggtaaa|tttaccct' },
    { regex: /aggggtaaaa|tttacccct/ig, name: 'aggggtaaaa|tttacccct' },
    { regex: /agggggtaaaaa|tttaacccct/ig, name: 'agggggtaaaaa|tttaacccct' },
    { regex: /[cgt]gggtaaa|tttaccc[acg]/ig, name: '[cgt]gggtaaa|tttaccc[acg]' },
    { regex: /a[act]ggtaaa|tttacc[agt]t/ig, name: 'a[act]ggtaaa|tttacc[agt]t' },
    { regex: /ag[act]gtaaa|tttac[agt]ct/ig, name: 'ag[act]gtaaa|tttac[agt]ct' },
    { regex: /agg[act]taaa|ttta[agt]cct/ig, name: 'agg[act]taaa|ttta[agt]cct' },
    { regex: /aggg[acg]aaa|ttt[cgt]ccct/ig, name: 'aggg[acg]aaa|ttt[cgt]ccct' },
    { regex: /agggt[cgt]aa|tt[acg]accct/ig, name: 'agggt[cgt]aa|tt[acg]accct' },
    { regex: /agggta[cgt]a|t[acg]taccct/ig, name: 'agggta[cgt]a|t[acg]taccct' },
    { regex: /agggtaa[cgt]|[acg]ttaccct/ig, name: 'agggtaa[cgt]|[acg]ttaccct' }
  ];
  
  for (const pattern of patterns) {
    const matches = sequence.match(pattern.regex) || [];
    console.log(`${pattern.name} ${matches.length}`);
  }
  
  // Perform replacements
  let result = sequence;
  result = result.replace(/tHa[Nt]/g, '<4>');
  result = result.replace(/aND|caN|Ha[DS]|WaS/g, '<3>');
  result = result.replace(/a[NSt]|BY/g, '<2>');
  result = result.replace(/<[^>]*>/g, '|');
  result = result.replace(/\|[^|][^|]*\|/g, '-');
  
  console.log();
  console.log(initialLength);
  console.log(codeLength);
  console.log(result.length);
}

main();

File: benchmarks/regex-redux/regex_redux.rs
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
//
// regex-redux benchmark in Rust

use std::io::{self, Read};
use regex::Regex;

fn main() {
    // Read input
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    
    // Remove header and newlines
    let sequence = Regex::new(r">.*\n|\n").unwrap().replace_all(&input, "");
    
    // Count patterns
    let initial_length = input.len();
    let code_length = sequence.len();
    
    let patterns = [
        (r"(?i)agggtaaa|tttaccct", "agggtaaa|tttaccct"),
        (r"(?i)aggggtaaaa|tttacccct", "aggggtaaaa|tttacccct"),
        (r"(?i)agggggtaaaaa|tttaacccct", "agggggtaaaaa|tttaacccct"),
        (r"(?i)[cgt]gggtaaa|tttaccc[acg]", "[cgt]gggtaaa|tttaccc[acg]"),
        (r"(?i)a[act]ggtaaa|tttacc[agt]t", "a[act]ggtaaa|tttacc[agt]t"),
        (r"(?i)ag[act]gtaaa|tttac[agt]ct", "ag[act]gtaaa|tttac[agt]ct"),
        (r"(?i)agg[act]taaa|ttta[agt]cct", "agg[act]taaa|ttta[agt]cct"),
        (r"(?i)aggg[acg]aaa|ttt[cgt]ccct", "aggg[acg]aaa|ttt[cgt]ccct"),
        (r"(?i)agggt[cgt]aa|tt[acg]accct", "agggt[cgt]aa|tt[acg]accct"),
        (r"(?i)agggta[cgt]a|t[acg]taccct", "agggta[cgt]a|t[acg]taccct"),
        (r"(?i)agggtaa[cgt]|[acg]ttaccct", "agggtaa[cgt]|[acg]ttaccct"),
    ];
    
    for (pattern, name) in &patterns {
        let re = Regex::new(pattern).unwrap();
        let count = re.find_iter(&sequence).count();
        println!("{} {}", name, count);
    }
    
    // Perform replacements
    let result = Regex::new(r"tHa[Nt]").unwrap().replace_all(&sequence, "<4>");
    let result = Regex::new(r"aND|caN|Ha[DS]|WaS").unwrap().replace_all(&result, "<3>");
    let result = Regex::new(r"a[NSt]|BY").unwrap().replace_all(&result, "<2>");
    let result = Regex::new(r"<[^>]*>").unwrap().replace_all(&result, "|");
    let result = Regex::new(r"\|[^|][^|]*\|").unwrap().replace_all(&result, "-");
    
    println!();
    println!("{}", initial_length);
    println!("{}", code_length);
    println!("{}", result.len());
}
